openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,323 @@
+// resource
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister   corev1listers.NamespaceLister
+	nodeLister corev1listers.NodeLister
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace %q: %v", PluginName, attr.GetNamespace(), err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isNamespaceAllowed(ns.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	if !isPodManaged(pod.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving nodes: %v", PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isManagementResourceAvailable(nodes) {
+		klog.Warningf("%s no nodes with resource %s is available", PluginName, managedCapacityLabel)
+
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+		return nil
+	}
+
+	allContainers := append([]coreapi.Container{}, pod.Spec.InitContainers...)
+	allContainers = append(allContainers, pod.Spec.Containers...)
+	podQoSClass := getPodQoSClass(allContainers)
+
+	// we do not want to change guaranteed pods resource allocation, because it should be managed by
+	// relevant managers(CPU and memory) under the kubelet
+	if podQoSClass == coreapi.PodQOSGuaranteed {
+		klog.Warningf("Skip pod %q CPUs requests modifications under namespace %q because it has guaranteed QoS class", pod.Name, pod.Namespace)
```

Comment:
why a warning? What does this want for?

Reply:
If we want to signal the owner of the pod, we should do something to the pod itself. Maybe an event? Or another annotation?

Reply:
I thought mostly about debugability if the pod passed namespace and managed check I want to see why it failed to apply changes, I can move it to something like
`klog.V(5).Infof(...)`

Reply:
Regular users can't see the logs, though, right? So we need to expose the error in some other way.

Reply:
@sttts Does it possible to use event API under the admission plugin, or does it a bad practice?
Probably annotation is the only option to inform a user.